test/mail-test-srv/http.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+)
+
+// filter is an interface to filter the received mails based on some criteria
+type filter interface {
+	Match(rcvdMail) bool
+}
+
+// toFromFilter filters mails based on the To: and From: fields.
+// The zero value matches all mails.
+type toFromFilter struct {
+	From string
+	To   string
+}
+
+func (f *toFromFilter) Match(m rcvdMail) bool {
+	if f.From != "" && f.From != m.From {
+		return false
+	}
+	if f.To != "" && f.To != m.To {
+		return false
+	}
+	return true
+}
+
+/*
+/count - number of mails
+/count?to=foo@bar.com - number of mails for foo@bar.com
+/count?from=foo@bar.com - number of mails from foo@bar.com
+/clear - clear the mail list
+/mail/0 - first mail
+/mail/1 - second mail
+/mail/0?to=foo@bar.com - first mail for foo@bar.com
+*/
+
+func setupHTTP(serveMux *http.ServeMux) {
+	serveMux.HandleFunc("/count", httpCount)
+	serveMux.HandleFunc("/clear", httpClear)
+	serveMux.HandleFunc("/mail/", httpGetMail)
+}
+
+func httpClear(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "POST" {
+		allMailMutex.Lock()
+		allReceivedMail = nil
+		allMailMutex.Unlock()
+		w.WriteHeader(200)
+	} else {
+		w.WriteHeader(405)
+	}
+}
+
+func httpCount(w http.ResponseWriter, r *http.Request) {
+	count := 0
+	iterMail(extractFilter(r), func(m rcvdMail) bool {
+		count++
+		return false
+	})
+	fmt.Fprintf(w, "%d\n", count)
+}
+
+var rgxGetMail = regexp.MustCompile(`^/mail/(\d+)/?$`)
```

Comment:
Why allow getting mail by offset, rather than just splatting out a JSON list of all the mails received?